rust/kernel/module_param.rs

```diff
@@ -0,0 +1,67 @@
+use core::fmt::Write;
+
+/// Types that can be used for module parameters.
+/// Note that displaying the type in `sysfs` will fail if `to_string` returns
+/// more than 4K bytes (including an additional null terminator).
+pub trait ModuleParam : core::fmt::Display + core::marker::Sized {
+    fn try_from_param_arg(arg: &[u8]) -> Option<Self>;
+
+    /// # Safety
+    ///
+    /// `val` must point to a valid null-terminated string. The `arg` field of
+    /// `param` must be an instance of `Self`.
+    unsafe extern "C" fn set_param(val: *const crate::c_types::c_char, param: *const crate::bindings::kernel_param) -> crate::c_types::c_int {
+        let arg = crate::c_types::c_string_bytes(val);
+        match Self::try_from_param_arg(arg) {
+            Some(new_value) => {
+                let old_value = (*param).__bindgen_anon_1.arg as *mut Self;
+                let _ = core::ptr::replace(old_value, new_value);
+                0
+            }
+            None => crate::error::Error::EINVAL.to_kernel_errno()
+        }
+    }
+
+    /// # Safety
+    ///
+    /// `buf` must be a buffer of length at least `kernel::PAGE_SIZE` that is
+    /// writeable. The `arg` field of `param` must be an instance of `Self`.
+    unsafe extern "C" fn get_param(buf: *mut crate::c_types::c_char, param: *const crate::bindings::kernel_param) -> crate::c_types::c_int {
+        let slice = core::slice::from_raw_parts_mut(buf as *mut u8, crate::PAGE_SIZE);
+        let mut buf = crate::buffer::Buffer::new(slice);
+        match write!(buf, "{}\0", *((*param).__bindgen_anon_1.arg as *mut Self)) {
+            Err(_) => crate::error::Error::EINVAL.to_kernel_errno(),
+            Ok(()) => buf.bytes_written() as crate::c_types::c_int,
+        }
+    }
+
+    /// # Safety
+    ///
+    /// The `arg` field of `param` must be an instance of `Self`.
+    unsafe extern "C" fn free(arg: *mut crate::c_types::c_void) {
+        core::ptr::drop_in_place(arg as *mut Self);
+    }
+}
+
+macro_rules! make_param_ops {
+    ($ops:ident, $ty:ident) => {
+        impl ModuleParam for $ty {
+            fn try_from_param_arg(arg: &[u8]) -> Option<Self> {
+                let utf8 = core::str::from_utf8(arg).ok()?;
+                utf8.parse::<$ty>().ok()
+            }
+        }
+
+        pub static $ops: crate::bindings::kernel_param_ops = crate::bindings::kernel_param_ops {
+            flags: 0,
+            set: Some(<$ty as crate::module_param::ModuleParam>::set_param),    
+            get: Some(<$ty as crate::module_param::ModuleParam>::get_param),
+            free: Some(<$ty as crate::module_param::ModuleParam>::free),
+        };
```

Comment:
If we are doing parsing in our side for some of the types, I am wondering if we should do it for all our types, i.e. just always use `charp`. Thoughts?

Reply:
The downsides?

I think if we use `charp` for everything then we'll be doing allocations for types that don't need it and we'd have to add some extra logic if we don't want to re-parse every time the param is read by the module. For those reasons I think using our own `ops` would be better.

Seems like using our own ops for everything would be fine, but we should make sure that we can at least parse everything that's supported for the existing params. For bool we'd need to accept 'yYnN01' and for the integer types we'd need to add support for hex and octal (but maybe we should do that anyway for these types?).

Reply:
> for the integer types we'd need to add support for hex and octal (but maybe we should do that anyway for these types?)

Exactly, I was thinking of supporting hex, octal and also binary for all of them. The C side doesn't currently, because they just pass a particular format string to `scnprintf()` and I guess they didn't want to complicate the parsing for the others, so they added `hexint` independently...

Reply:
Long-term, the C side could call our safe parsing functions instead ;-)

Reply:
Added parsing of hex/octal/binary and switch to the Rust param ops for all integer types. Using our own param ops for `bool` will require adding support for the `NOARG` flag (which we'll want for custom params anyway) so I can do a separate PR for that and `charp`.

I tried to split the parsing logic into a separate crate that could be used both here and in the `module` proc macro, but it got a bit tricky: The common logic will need to be compiled for the host so that it can be used in the proc macro and also for the target architecture so that it can be used here. Any ideas on the best way to do that?

Also, just to make sure I'm not mixed up, I think the C side supports parsing hex and octal for all the int params (via `kstrtol`) and `hexint` is just different is that it displays as hex via it's `get` function?

Reply:
We can have a crate that is shared for both host and target -- I guess the need will come up sooner or later. I can take a look, although I am not sure if we should add it for the first PR.

For this particular case though, if we end up using `syn`, we wouldn't need it, since it could parse the literals. Another option, we add it to `core` ;-) There is https://github.com/rust-lang/rfcs/issues/1098 for that.

And you're 100% right, I inverted `get/set`, the format string is for the display side.

Reply:
On second thought it actually wasn't much of a change to use our own parser for `bool` so just added it here. I don't think we should do a `charp` replacement just now though. I think the inner type would need to be some type of enum so that we can distinguish between the cases where the param needs to freed or not. But then we'd run into the issue of the `default` field in the macro needing to be a literal. So that should probably wait until `syn` too.